### PR TITLE
Add missing methods to (de)serializer interfaces and implementations

### LINKED ISSUE #828: missing `IDeserializer object? Deserialize(string input)` method
**Is your feature request related to a problem? Please describe.**
Currently, there are matching pairs of methods on this interface, but one combination is missing:
```
    T Deserialize<T>(string input);
    T Deserialize<T>(TextReader input);
    T Deserialize<T>(IParser parser);

    // no object? Deserialize(string input); here :(
    object? Deserialize(TextReader input);
    object? Deserialize(IParser parser);

    object? Deserialize(string input, Type type);
    object? Deserialize(TextReader input, Type type);
    object? Deserialize(IParser parser, Type type); // the "real" method, i.e. the one that the others call in impl
```

It makes it impossible to call the deserializer this way.

**Describe the solution you'd like**
Add the missing interface method and implementation. See below for an example implementation for `Deserializer` class.

**Describe alternatives you've considered**
```
static class Extensions {
    public static object? Deserialize(this IDeserializer iSerializer, string input)
    {
        using (StringReader input1 = new StringReader(input))
            return iSerializer.Deserialize((TextReader) input1);
    }
}
```
or similar (in new .NET/C# versions, the body it can be even as simple as
```
        using StringReader input1 = new(input);
        return iSerializer.Deserialize(input1);
```
)

**Additional context**
In case of "sure, but no manpower, PR welcome", I can gladly do a PR for this as soon as I have some spare time, since it's trivial :D

### PR DESCRIPTION
**Pull request by FyiurAmron**
Add missing methods to (de)serializer interfaces and implementations - also reorder the methods to make a sorted sequence.

Closes #828

### CODE REVIEW

File: YamlDotNet/Serialization/Deserializer.cs [hunk 1]
```diff
@@ -77,11 +77,26 @@ public T Deserialize<T>(TextReader input)
             return Deserialize<T>(new Parser(input));
         }
 
+        public T Deserialize<T>(IParser parser)
+        {
+            return (T)Deserialize(parser, typeof(T))!; // We really want an exception if we are trying to deserialize null into a non-nullable type
```
Comment (EdwardCooke) [inline YamlDotNet/Serialization/Deserializer.cs:82]: As far as I'm aware, just adding an `!` to the line won't throw an exception. It just tells the compiler that it'll never be null to avoid the compiler warnings. Removing the comment may be warranted. And changing the resulting type to `T?` and removing the `!` since it is very possible deserializing could be null.

If `T` is a `struct` it wouldn't be a null value, but if it's a class it would be.

Comment (EdwardCooke) [reply YamlDotNet/Serialization/Deserializer.cs:82]: Nevermind, this is the same method that was already there. Looks like you just moved it in the code. So keep it as is.

**Comment by EdwardCooke** [comment]
Another issue here is that your branch shouldn't be on master, please create your own, otherwise the build will fail.

**Comment by FyiurAmron** [comment]
@EdwardCooke opened MR from a non-master branch , https://github.com/aaubry/YamlDotNet/pull/832

File: YamlDotNet/Serialization/Serializer.cs [hunk 2]
```diff
@@ -63,25 +63,37 @@ public static Serializer FromValueSerializer(IValueSerializer valueSerializer, E
...
+        /// <param name="type">The static type of the object to serialize.</param> 
+        public string Serialize(object? graph, Type type)
         {
             using var buffer = new StringWriter();
-            Serialize(buffer, graph);
+            Serialize(buffer, graph, type);
             return buffer.ToString();
         }
+        
+        /// <summary>
```
[the remark this answers is not in the document]
Comment (FyiurAmron) [reply YamlDotNet/Serialization/Serializer.cs:88]: @EdwardCooke it was reordered to match the interface reordering, so that the implementation has the same order as the interface definition

File: YamlDotNet/Serialization/ISerializer.cs [hunk 3]
```diff
@@ -28,17 +28,24 @@ namespace YamlDotNet.Serialization
     public interface ISerializer
     {
         /// <summary>
-        /// Serializes the specified object.
+        /// Serializes the specified object into a string.
```
[the remark this answers is not in the document]
Comment (FyiurAmron) [reply YamlDotNet/Serialization/ISerializer.cs:31]: yes, it was reordered to make a logical sequence of the defined method, in order to prevent similar mishaps/omissions in the future. 
```
        T Deserialize<T>(string input);
        T Deserialize<T>(TextReader input);
        T Deserialize<T>(IParser parser);

        object? Deserialize(string input);
        object? Deserialize(TextReader input);
        object? Deserialize(IParser parser);

        object? Deserialize(string input, Type type);
        object? Deserialize(TextReader input, Type type);
        object? Deserialize(IParser parser, Type type);
```
as you can see, the sequence now is basically
```
R1(P1)
R1(P2)
R1(P3)

R2(P1)
R2(P2)
R2(P3)

R2(P1,P4)
R2(P2,P4)
R2(P3,P4)
```
with the most concrete method being the last, and the helpers calling/referencing it defined before.

tl;dr Basically, it was sorted semi-lexicographically
